awfm9/flow-dps
# Add GRPC API

## Goal of this PR

This PR fixes #12 by implementing the GRPC API Server and its Controller. It also

* Adds a file that allows running `go generate` to rebuild the protobuf files
* Adds a GitHub action workflow that ensures that PR authors that made changes that impact the protobuf file generation did correctly include those changes in their PR
* Adds documentation regarding the newly introduced dependencies (binaries required for building protobuf files) in the README.md file at the root of the repository
* Moves the `rest.State` interfaces to `model` so that they can be shared between `state`, `rest` and `grpc`.

## How to test it

Currently there are no tests of any kind anywhere within the code. You cam however manually call the controller from `main.go` to check that it gets results as expected, or code your own small GRPC client for testing.

<details>
<summary>Example call in `main.go`</summary>

Make sure to replace the value key part used in the following eample with the relevant parts from your index.

```go
		valueKey, _ := hex.DecodeString("6163636f756e745f616464726573735f7374617465")
		vals, err := grpcController.GetValues(context.Background(), &grpcApi.GetValuesRequest{
			Hash:    nil,
			Version: nil,
			Keys:    []*grpcApi.Key{
				{
					Parts: []*grpcApi.KeyPart{
						{
							Type: 0,
							Value: []byte{},
						},
						{
							Type: 1,
							Value: []byte{},
						},
						{
							Type: 2,
							Value: valueKey,
						},
					},
				},
			},
		})
		if err != nil {
			log.Fatal().Err(err).Msg("unable to get values")
		}
		log.Info().Str("values", vals.String()).Send()
```

Expected output:

```json
{"level":"info","values":"values:\"\\x00\\x00\\x00\\x00\\x00C\"","time":"2021-05-11T07:49:40Z"}
```

For reference, the same call over the GRPC API provides the following answer:

```json
["000000000043"]
```

</details>

## Additional Notes

~This PR is a draft for now, as since it adds a GitHub action workflow, it can't be manually ran (AFAIK) until there's an open PR that includes it.~
